### PR TITLE
Improve perf of all buffer consuming APIs

### PR DESCRIPTION
**Pull request by SUPERCILEX**
Rustix being a low-level library, it should not have worse perf than the stdlib: https://github.com/rust-lang/rust/blob/0fb8b72ce49997d60a631e921d2cf5be9ca229e6/library/std/src/sys/unix/fs.rs#L1396-L1423

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
Turns out ttyname depends on readlink, so I just converted all the buffer consuming APIs to the new approach. `getcwd` from the stdlib: https://github.com/rust-lang/rust/blob/0fb8b72ce49997d60a631e921d2cf5be9ca229e6/library/std/src/sys/unix/os.rs#L142-L166

**Comment by SUPERCILEX** [comment]
Well shucks, all of this stuff requires Rust 1.60. I guess we can just let this sit for a few months until Bookworm is released? It currently has 1.63: https://packages.debian.org/testing/rust/rustc

File: src/backend/linux_raw/termios/syscalls.rs [hunk 1]
```diff
@@ -238,14 +239,19 @@ pub(crate) fn ttyname(fd: BorrowedFd<'_>, buf: &mut [u8]) -> io::Result<usize> {
...
     }
-    buf[r] = b'\0';
+    buf[r].write(b'\0');
 
     // Check that the path we read refers to the same file as `fd`.
-    let path = CStr::from_bytes_with_nul(&buf[..=r]).unwrap();
+    {
+        // SAFETY:
+        // - We just wrote the NUL byte above
+        // - readlinkat() returns the number of bytes placed in the buffer
```
Comment (sunfishcode) [inline src/backend/linux_raw/termios/syscalls.rs:248]: As a minor style nit, I prefer to see function names code-quoted and not suffixed with `()` unless they actually have zero arguments.  The old `()` convention reminds me too much of old-school K&R C where you could just `()` even if the function has args :-}.

**Review by sunfishcode — COMMENTED** [review]
Nice! And yeah, I'm really looking forward to bumping the MSRV too, as 1.48 is getting increasingly burdensome, but for our reverse dependencies it would be good to wait a little longer.

File: src/backend/linux_raw/termios/syscalls.rs (hunk 1 above)
Comment (SUPERCILEX) [reply src/backend/linux_raw/termios/syscalls.rs:248]: Lol, fixed.

**Comment by SUPERCILEX** [comment]
> for our reverse dependencies it would be good to wait a little longer

Sounds good!

**Comment by SUPERCILEX** [comment]
@sunfishcode Lemme know when master stabilizes and I'll resolve the conflicts here.

**Comment by sunfishcode** [comment]
@SUPERCILEX The 0.38-planning branch is now landed on main, with the MSRV bump, so we're now ready!

**Comment by sunfishcode** [comment]
Thanks!

**Pull request merged**